accepted/0000-hub_retail.md

```diff
@@ -0,0 +1,310 @@
+
+Alternative 2:
+Central creation of groups is covered in [Hub Users RFC](00061-hub-users.md). The salt states
+can be extended to handle also formula data.
+This approach is probably more flexible.
+
+Both alternatives will have to cope with authentization and Hub vs. Peripheral server user credentials.
+
+In long term, alternative 2 is clean solution.
+
+Alternative 1 might be used as a short term fix and also to keep compatibility.
+
+Note:
+The Formula pillar files depend on group IDs. The IDs differs between Hub and Peripheral server so
+direct file-based synchronization of pillars is not possible.
+
+
+YAML configuration
+------------------
+
+FIXME: optional, related to Alternative 2 in previous section
+
+We have retail_yaml tool for initial configuration of Retail environment from
+[YAML](https://github.com/uyuni-project/retail/blob/master/python-susemanager-retail/example.yml) file.
+[Hub Users RFC](00061-hub-users.md) introduced similar concept, based on Salt.
+Maybe these parts could be merged.
+
+
+Combined Peripheral and Branch server
+-------------------------------------
+
+Branch services are configured by Salt formulas. Also PXE entries for each terminal are generated
+by Salt (currently via reactor).
+With combined Peripheral and Branch server, these formulas should be applied on SUMA server itself.
+
+Terminals connect to SUMA directly, SUMA Proxy is not needed.
+
+
+FIXME: pick one alternative
+
+Alternative 1: Hub as a Salt Master for Combined Peripheral and Branch server
+
+- this is already used for [Hub Users RFC](00061-hub-users.md)
+
+problems:
+- does not work for PXE entries https://github.com/uyuni-project/uyuni-rfc/pull/46
+  The event from each booting Terminal must be stopped on Peripheral server, on
+  Hub it would cause the same scalability problems as before.
+
+- mixed environment with Hub, Peripheral Servers with standalone Branch Serveres and
+  Combined Peripheral Servers will have non-optimal configuration
+  For example: A shop uses Combined Peripheral and Branch server, then the number of terminals grows and
+  at some point it is necessary to add standalone Branch Server. Then the configuration of the
+  shop would be split between Hub and Peripheral Server, or the shop would have to be reinstalled
+  with 2 Branch Servers and standalone Peripheral Server
+
+- does not cover the "Single server" setup
+
+
+Alternative 2: Multi-master setup of Combined Peripheral and Branch server
+
+- Salt minion running on Combined Peripheral and Branch server has 2 masters - Hub and itself
+- the Branch Server configuration via formulas can stay unchanged
+
+problems:
+- SUMA registered to itself cause many problems
+  - we already have clients that do not allow certain actions (containers) so this should be fixable
+  - if we allow and test this scenario for Retail, we will know how it behaves for
+  non-Retail use and the most severe problems will be fixed (instead of just saying "we don't know")
```

Comment:
A few problems are known, and I can explain them.

Most important is the Server package upgrade problem. Registering a minion against a Server implies deactivating all of the normal, plain http zypper repos that are configured and replacing them with the Server's assigned channels. Those are served dynamically via Java code.

Problem is: once a Server is registered to itself, how to upgrade it? In order to upgrade Server packages you have to take Server services down first, and at that point your repos will not be available any more. One option would be to keep plain http repos around or the Server won't be upgradable. This is hairy but could be fixed in principle.

If you do that, you will bump into the next problem: the availability of updates on Server would not be displayed correctly. That follows because it's only possible if the repos are really channels, and not plain http repos. That's why we generally disable those upon registration/highstate, and I do not have a good solution in mind for now.

At leas when talking about the Salt stack we are willing to improve it, but we get another problem considering the traditional stack - there we do not want to add any new code per policy. And we know that the traditional stack completely breaks down if a Server is registered onto itself. So we would also need to selectively disallow traditional and allow only Salt for self-registration only.

Then we get to fix subscription matching, as it would count one extra subscription for the minion it currently expects not to be there.

And then, of course, we will have to make sure that no future configuration modification to minions interferes with the master or Server code, which is unexplored territory at this point.

As you can see it's not trivial, and I would be hesitant in considering this use case important enough to change the design - unless it becomes terribly complicated or fragile to do otherwise of course.